File: user-service/src/main/java/rs/raf/user_service/controller/ActuaryController.java
package rs.raf.user_service.controller;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.AllArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;
import rs.raf.user_service.domain.dto.*;
import rs.raf.user_service.exceptions.ActuaryLimitNotFoundException;
import rs.raf.user_service.exceptions.EmployeeNotFoundException;
import rs.raf.user_service.exceptions.UserNotAgentException;
import rs.raf.user_service.service.ActuaryService;

import javax.validation.Valid;

@RestController
@RequestMapping("/api/admin/actuaries")
@Tag(name = "Actuary Management", description = "API for managing actuaries")
@AllArgsConstructor
public class ActuaryController {

    private final ActuaryService actuaryService;

    @PreAuthorize("hasRole('SUPERVISOR')")
    @PutMapping("change-limit/{id}")
    @Operation(summary = "Change agent limit.")
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "Agent limit changed successfully."),
            @ApiResponse(responseCode = "404", description = "Agent not found"),
            @ApiResponse(responseCode = "400", description = "Invalid input data")
    })
    public ResponseEntity<?> changeAgentLimit(@PathVariable Long id, @Valid @RequestBody ChangeAgentLimitDto changeAgentLimitDto) {
        try {
            actuaryService.changeAgentLimit(id, changeAgentLimitDto.getNewLimit());
            return ResponseEntity.ok().build();
        } catch (ActuaryLimitNotFoundException | EmployeeNotFoundException | UserNotAgentException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.getMessage());
        }
    }

    @PreAuthorize("hasRole('SUPERVISOR')")
    @PutMapping("reset-limit/{id}")
    @Operation(summary = "Reset daily limit for an agent.")
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "Agent daily limit reset successfully."),
            @ApiResponse(responseCode = "400", description = "Invalid input data")
    })
    public ResponseEntity<?> resetDailyLimit(@PathVariable Long id) {
        try {
            actuaryService.resetDailyLimit(id);
            return ResponseEntity.ok().build();
        } catch (ActuaryLimitNotFoundException | EmployeeNotFoundException | UserNotAgentException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.getMessage());
        }
    }

    @PreAuthorize("hasRole('SUPERVISOR')")
    @PutMapping("set-approval/{id}")
    @Operation(summary = "Set approval value for an agent.")
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "Agent approval value set successfully."),
            @ApiResponse(responseCode = "400", description = "Invalid input data")
    })
    public ResponseEntity<?> setApprovalValue(@PathVariable Long id, @Valid @RequestBody SetApprovalDto setApprovalDto) {
        try {
            actuaryService.setApproval(id, setApprovalDto.getNeedApproval());
            return ResponseEntity.ok().build();
        } catch (ActuaryLimitNotFoundException | EmployeeNotFoundException | UserNotAgentException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.getMessage());
        }
    }

    @PreAuthorize("hasRole('SUPERVISOR')")
    @GetMapping("/agents")
    @Operation(summary = "Get all agents with filtering.")
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "Agents retrieved successfully")
    })
    public ResponseEntity<Page<EmployeeDto>> getAllAgents(
            @RequestParam(required = false) String email,
            @RequestParam(required = false) String firstName,
            @RequestParam(required = false) String lastName,
            @RequestParam(required = false) String position,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size) {
        Pageable pageable = PageRequest.of(page, size);
        return ResponseEntity.ok(actuaryService.findAgents(firstName, lastName, email, position, pageable));
    }

    @PreAuthorize("hasRole('SUPERVISOR')")
    @GetMapping
    @Operation(summary = "Get all actuaries.")
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "Actuaries retrieved successfully")
    })
    public ResponseEntity<Page<ActuaryDto>> getAllActuaries(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size) {
        Pageable pageable = PageRequest.of(page, size);
        return ResponseEntity.ok(actuaryService.findActuaries(pageable));
    }

    @PreAuthorize("hasAnyRole('SUPERVISOR','AGENT')")
    @GetMapping("{agentId}")
    @Operation(summary = "Get agent actuary limit.")
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "Agent actuary limit returned successfully."),
            @ApiResponse(responseCode = "404", description = "Not found.")
    })
    public ResponseEntity<?> getAgentLimit(@PathVariable Long agentId) {
        try {
            return ResponseEntity.ok().body(actuaryService.getAgentLimit(agentId));
        } catch (ActuaryLimitNotFoundException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(e.getMessage());
        }
    }
    @PreAuthorize("hasRole('ADMIN')")
    @GetMapping("/all")
    public ResponseEntity<?> getAllAgentsAndClients(
            @RequestParam(defaultValue = "") String name,
            @RequestParam(defaultValue = "") String surname,
            @RequestParam(defaultValue = "") String  role
    ){
        return ResponseEntity.ok().body(actuaryService.getAllAgentsAndClients(name,surname,role));
    }
}

File: user-service-kotlin/src/main/kotlin/pack/userservicekotlin/controller/ActuaryController.kt
package pack.userservicekotlin.controller

import jakarta.validation.Valid
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import pack.userservicekotlin.arrow.ActuaryServiceError
import pack.userservicekotlin.domain.dto.activity_code.SetApprovalDto
import pack.userservicekotlin.domain.dto.actuary_limit.ActuaryResponseDto
import pack.userservicekotlin.domain.dto.actuary_limit.UpdateActuaryLimitDto
import pack.userservicekotlin.domain.dto.employee.AgentDto
import pack.userservicekotlin.service.ActuaryService
import pack.userservicekotlin.swagger.ActuaryApiDoc

@RestController
@RequestMapping("/api/admin/actuaries")
class ActuaryController(
    private val actuaryService: ActuaryService,
) : ActuaryApiDoc {
    @PreAuthorize("hasRole('SUPERVISOR')")
    @PutMapping("change-limit/{id}")
    override fun changeAgentLimit(
        @PathVariable id: Long,
        @Valid @RequestBody changeActuaryLimitDto: UpdateActuaryLimitDto,
    ): ResponseEntity<Any> =
        actuaryService.changeAgentLimit(id, changeActuaryLimitDto.newLimit).fold(
            ifLeft = {
                when (it) {
                    is ActuaryServiceError.EmployeeNotFound ->
                        ResponseEntity.status(HttpStatus.NOT_FOUND).body("Employee with id ${it.id} not found")
                    is ActuaryServiceError.NotAnAgent ->
                        ResponseEntity.status(HttpStatus.BAD_REQUEST).body("User with id ${it.employeeId} is not an agent")
                    is ActuaryServiceError.ActuaryLimitNotFound ->
                        ResponseEntity.status(HttpStatus.NOT_FOUND).body("Limit not found for agent ${it.employeeId}")
                    is ActuaryServiceError.ExternalServiceError -> TODO()
                    ActuaryServiceError.InvalidPageRequest -> TODO()
                }
            },
            ifRight = { ResponseEntity.ok().build() },
        )

    @PreAuthorize("hasRole('SUPERVISOR')")
    @PutMapping("reset-limit/{id}")
    override fun resetDailyLimit(
        @PathVariable id: Long,
    ): ResponseEntity<Any> =
        actuaryService.resetDailyLimit(id).fold(
            ifLeft = {
                when (it) {
                    is ActuaryServiceError.EmployeeNotFound ->
                        ResponseEntity.status(HttpStatus.NOT_FOUND).body("Employee with id ${it.id} not found")
                    is ActuaryServiceError.NotAnAgent ->
                        ResponseEntity.status(HttpStatus.BAD_REQUEST).body("User with id ${it.employeeId} is not an agent")
                    is ActuaryServiceError.ActuaryLimitNotFound ->
                        ResponseEntity.status(HttpStatus.NOT_FOUND).body("Limit not found for agent ${it.employeeId}")

                    is ActuaryServiceError.ExternalServiceError -> TODO()
                    ActuaryServiceError.InvalidPageRequest -> TODO()
                }
            },
            ifRight = { ResponseEntity.ok().build() },
        )

    @PreAuthorize("hasRole('SUPERVISOR')")
    @PutMapping("set-approval/{id}")
    override fun setApprovalValue(
        @PathVariable id: Long,
        @Valid @RequestBody setApprovalDto: SetApprovalDto,
    ): ResponseEntity<Any> =
        actuaryService.setApproval(id, setApprovalDto.needApproval!!).fold(
            ifLeft = {
                when (it) {
                    is ActuaryServiceError.EmployeeNotFound ->
                        ResponseEntity.status(HttpStatus.NOT_FOUND).body("Employee with id ${it.id} not found")
                    is ActuaryServiceError.NotAnAgent ->
                        ResponseEntity.status(HttpStatus.BAD_REQUEST).body("User with id ${it.employeeId} is not an agent")
                    is ActuaryServiceError.ActuaryLimitNotFound ->
                        ResponseEntity.status(HttpStatus.NOT_FOUND).body("Limit not found for agent ${it.employeeId}")

                    is ActuaryServiceError.ExternalServiceError -> TODO()
                    ActuaryServiceError.InvalidPageRequest -> TODO()
                }
            },
            ifRight = { ResponseEntity.ok().build() },
        )

    @PreAuthorize("hasAnyRole('SUPERVISOR','AGENT')")
    @GetMapping("{agentId}")
    override fun getAgentLimit(
        @PathVariable agentId: Long,
    ): ResponseEntity<Any> =
        actuaryService.getAgentLimit(agentId).fold(
            ifLeft = {
                when (it) {
                    is ActuaryServiceError.ActuaryLimitNotFound ->
                        ResponseEntity.status(HttpStatus.NOT_FOUND).body("Limit not found for agent ${it.employeeId}")
                    else -> ResponseEntity.internalServerError().build()
                }
            },
            ifRight = { ResponseEntity.ok(it) },
        )

    @PreAuthorize("hasRole('SUPERVISOR')")
    @PutMapping("update-used-limit/{id}")
    override fun updateUsedLimit(
        @PathVariable id: Long,
        @Valid @RequestBody changeActuaryLimitDto: UpdateActuaryLimitDto,
    ): ResponseEntity<*> =
        actuaryService.updateUsedLimit(id, changeActuaryLimitDto.newLimit).fold(
            ifLeft = { error ->
                when (error) {
                    is ActuaryServiceError.ActuaryLimitNotFound,
                    is ActuaryServiceError.EmployeeNotFound,
                    is ActuaryServiceError.NotAnAgent,
                    -> ResponseEntity.status(HttpStatus.BAD_REQUEST).body(error)
                    is ActuaryServiceError.ExternalServiceError -> TODO()
                    ActuaryServiceError.InvalidPageRequest -> TODO()
                }
            },
            ifRight = { dto -> ResponseEntity.ok(dto) },
        )

    @PreAuthorize("hasRole('SUPERVISOR')")
    @GetMapping("/agents")
    override fun getAllAgents(
        @RequestParam(required = false) email: String?,
        @RequestParam(required = false) firstName: String?,
        @RequestParam(required = false) lastName: String?,
        @RequestParam(required = false) position: String?,
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "10") size: Int,
    ): ResponseEntity<Page<AgentDto>> {
        val pageable = PageRequest.of(page, size)
        return actuaryService.findAgents(firstName, lastName, email, position, pageable).fold(
            ifLeft = {
                when (it) {
                    is ActuaryServiceError.ActuaryLimitNotFound ->
                        ResponseEntity
                            .status(HttpStatus.NOT_FOUND)
                            .body(PageImpl(emptyList())) // could log or include message
                    else -> ResponseEntity.internalServerError().build()
                }
            },
            ifRight = { ResponseEntity.ok(it) },
        )
    }

    @PreAuthorize("hasRole('SUPERVISOR')")
    @GetMapping
    override fun getAllActuaries(
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "10") size: Int,
    ): ResponseEntity<Page<ActuaryResponseDto>> {
        val pageable: Pageable = PageRequest.of(page, size)
        return actuaryService.findActuaries(pageable).fold(
            ifLeft = { ResponseEntity.internalServerError().build() },
            ifRight = { ResponseEntity.ok(it) },
        )
    }

    @PreAuthorize("hasRole('ADMIN')")
    @GetMapping("/all")
    override fun getAllAgentsAndClients(
        @RequestParam(defaultValue = "") name: String,
        @RequestParam(defaultValue = "") surname: String,
        @RequestParam(defaultValue = "") role: String,
    ): ResponseEntity<Any> =
        actuaryService.getAllAgentsAndClients(name, surname, role).fold(
            ifLeft = { ResponseEntity.internalServerError().body("Failed to fetch agents and clients") },
            ifRight = { ResponseEntity.ok(it) },
        )
}
